Can't have empty edata for graph with no edges

Consider:

Without any `edata`:
```julia-repl
julia> using GNNGraphs

julia> GNNGraph((Int[], Int[]); ndata=(;a=[1]),num_nodes=1)
GNNGraph:
  num_nodes: 1
  num_edges: 0
  ndata:
    a = 1-element Vector{Int64}
```

But if we have `edata=(;b=[])`:
```julia-repl
julia> GNNGraph((Int[], Int[]); ndata=(;a=[1]), edata=(;b=[]), num_nodes=1)
ERROR: AssertionError: DataStore: data[b] has 1 observations, but n = 0
Stacktrace:
 [1] DataStore(n::Int64, data::Dict{Symbol, Any})
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/datastore.jl:66
 [2] DataStore(n::Int64, data::@NamedTuple{b::Matrix{Any}})
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/datastore.jl:74
 [3] normalize_graphdata(data::@NamedTuple{b::Vector{Any}}; default_name::Symbol, n::Int64, duplicate_if_needed::Bool)
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/utils.jl:169
 [4] normalize_graphdata
   @ ~/.julia/packages/GNNGraphs/qLIkY/src/utils.jl:132 [inlined]
 [5] GNNGraph(data::Tuple{…}; num_nodes::Int64, graph_indicator::Nothing, graph_type::Symbol, dir::Symbol, ndata::@NamedTuple{…}, edata::@NamedTuple{…}, gdata::Nothing)
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/gnngraph.jl:141
 [6] GNNGraph
   @ ~/.julia/packages/GNNGraphs/qLIkY/src/gnngraph.jl:119 [inlined]
 [7] #GNNGraph#22
   @ ~/.julia/packages/GNNGraphs/qLIkY/src/gnngraph.jl:167 [inlined]
 [8] top-level scope
   @ REPL[68]:1
Some type information was truncated. Use `show(err)` to see complete types.
```
Or if we have `edata=[]`
```julia-repl
ulia> GNNGraph((Int[], Int[]); ndata=(;a=[1]), edata=[], num_nodes=1)
ERROR: AssertionError: DataStore: data[e] has 1 observations, but n = 0
Stacktrace:
 [1] DataStore(n::Int64, data::Dict{Symbol, Any})
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/datastore.jl:66
 [2] DataStore(n::Int64, data::@NamedTuple{e::Matrix{Any}})
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/datastore.jl:74
 [3] normalize_graphdata(data::@NamedTuple{e::Vector{Any}}; default_name::Symbol, n::Int64, duplicate_if_needed::Bool)
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/utils.jl:169
 [4] normalize_graphdata(data::Vector{Any}; default_name::Symbol, kws::@Kwargs{n::Int64, duplicate_if_needed::Bool})
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/utils.jl:129
 [5] normalize_graphdata
   @ ~/.julia/packages/GNNGraphs/qLIkY/src/utils.jl:128 [inlined]
 [6] GNNGraph(data::Tuple{…}; num_nodes::Int64, graph_indicator::Nothing, graph_type::Symbol, dir::Symbol, ndata::@NamedTuple{…}, edata::Vector{…}, gdata::Nothing)
   @ GNNGraphs ~/.julia/packages/GNNGraphs/qLIkY/src/gnngraph.jl:141
 [7] GNNGraph
   @ ~/.julia/packages/GNNGraphs/qLIkY/src/gnngraph.jl:119 [inlined]
 [8] #GNNGraph#22
   @ ~/.julia/packages/GNNGraphs/qLIkY/src/gnngraph.jl:167 [inlined]
 [9] top-level scope
   @ REPL[70]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

My workflow initializes the edata like that then pushes stuff to it as it processes the data i am generating a graph from.

--- a/GNNGraphs/src/gnngraph.jl
+++ b/GNNGraphs/src/gnngraph.jl
@@ -146,11 +146,9 @@ function GNNGraph(data::D;
     edata = normalize_graphdata(edata, default_name = :e, n = num_edges,
                                 duplicate_if_needed = true)
 
-    # don't force the shape of the data when there is only one graph
-    gdata = normalize_graphdata(gdata, default_name = :u,
-                                n = num_graphs > 1 ? num_graphs : -1)
+    gdata = normalize_graphdata(gdata, default_name = :u, n = num_graphs, glob=true)
 
-    GNNGraph(graph,
+    return GNNGraph(graph,
              num_nodes, num_edges, num_graphs,
              graph_indicator,
              ndata, edata, gdata)
@@ -203,7 +201,7 @@ function GNNGraph(g::GNNGraph; ndata = g.ndata, edata = g.edata, gdata = g.gdata
     ndata = normalize_graphdata(ndata, default_name = :x, n = g.num_nodes)
     edata = normalize_graphdata(edata, default_name = :e, n = g.num_edges,
                                 duplicate_if_needed = true)
-    gdata = normalize_graphdata(gdata, default_name = :u, n = g.num_graphs)
+    gdata = normalize_graphdata(gdata, default_name = :u, n = g.num_graphs, glob=true)
 
     if !isnothing(graph_type)
         if graph_type == :coo

--- a/GNNGraphs/src/gnnheterograph/gnnheterograph.jl
+++ b/GNNGraphs/src/gnnheterograph/gnnheterograph.jl
@@ -144,8 +144,7 @@ function GNNHeteroGraph(data::EDict;
         ndata = normalize_heterographdata(ndata, default_name = :x, ns = num_nodes)
         edata = normalize_heterographdata(edata, default_name = :e, ns = num_edges,
                                           duplicate_if_needed = true)
-        gdata = normalize_graphdata(gdata, default_name = :u,
-                                        n = num_graphs > 1 ? num_graphs : -1)
+        gdata = normalize_graphdata(gdata, default_name = :u, n = num_graphs, glob = true)
     end
 
     return GNNHeteroGraph(graph,

--- a/GNNGraphs/src/utils.jl
+++ b/GNNGraphs/src/utils.jl
@@ -129,19 +129,24 @@ function normalize_graphdata(data; default_name::Symbol, kws...)
     normalize_graphdata(NamedTuple{(default_name,)}((data,)); default_name, kws...)
 end
 
-function normalize_graphdata(data::NamedTuple; default_name, n, duplicate_if_needed = false)
+function normalize_graphdata(data::NamedTuple; default_name::Symbol, n::Int, 
+        duplicate_if_needed::Bool = false, glob::Bool = false)
     # This had to workaround two Zygote bugs with NamedTuples
     # https://github.com/FluxML/Zygote.jl/issues/1071
-    # https://github.com/FluxML/Zygote.jl/issues/1072
+    # https://github.com/FluxML/Zygote.jl/issues/1072 # TODO fixed. Can we simplify something?
+
 
     if n > 1
         @assert all(x -> x isa AbstractArray, data) "Non-array features provided."
     end
 
-    if n <= 1
-        # If last array dimension is not 1, add a new dimension.
-        # This is mostly useful to reshape global feature vectors
-        # of size D to Dx1 matrices.
+    if n <= 1 && glob == true
+        @assert n == 1
+        n = -1 # relax the case of a single graph, allowing to store arbitrary types
+    #     # # If last array dimension is not 1, add a new dimension.
+    #     # # This is mostly useful to reshape global feature vectors
+    #     # # of size D to Dx1 matrices.
+        # TODO remove this and handle better the batching of global features
         unsqz_last(v::AbstractArray) = size(v)[end] != 1 ? reshape(v, size(v)..., 1) : v
         unsqz_last(v) = v
 
@@ -161,7 +166,7 @@ function normalize_graphdata(data::NamedTuple; default_name, n, duplicate_if_nee
 
         for x in data
             if x isa AbstractArray
-                @assert size(x)[end]==n "Wrong size in last dimension for feature array, expected $n but got $(size(x)[end])."
+                @assert size(x)[end] == n "Wrong size in last dimension for feature array, expected $n but got $(size(x)[end])."
             end
         end
     end

--- a/GNNGraphs/test/gnngraph.jl
+++ b/GNNGraphs/test/gnngraph.jl
@@ -228,6 +228,17 @@ end
     @test g.num_nodes == 1
     @test g.num_edges == 0
     @test g.ndata.a == [1]
+
+    g = GNNGraph((Int[], Int[]); ndata=(;a=[1]), edata=(;b=Int[]), num_nodes=1)
+    @test g.num_nodes == 1
+    @test g.num_edges == 0
+    @test g.ndata.a == [1]
+    @test g.edata.b == Int[]
+
+    g = GNNGraph(; edata=(;b=Int[]))
+    @test g.num_nodes == 0
+    @test g.num_edges == 0
+    @test g.edata.b == Int[]
 end
 
 

